pysatModelUtils/utils/extract.py

```diff
@@ -24,9 +24,152 @@
+                                          scale=100.,
+                                          obs_out='model_altitude',
+                                          tol=1.,
+                                          autoscale=False):
+    """Interpolates altitude values onto model pressure levels.
+
+    Uses a recursive regular grid interpolation to find the
+    appropriate pressure level for the given input locations.
+
+    Parameters
+    ----------
+    obs : pysat.Instrument object
+        Instrument object with observational data
+    mod : pysat.Instrument object
+        Instrument object with modelled data
+    obs_coords : array-like
+        List of variable names containing the observational data coordinates
+        at which the model data will be interpolated. Must be in the same order
+        as the model dimensions.
+    obs_alt : string
+        String identifier used in inst for the altitude variable
+    mod_alt : string
+        Variable identifier for altitude data in the model
+        e.g. 'ZG' in standard TIEGCM files.
+    mod_coord : string
+        Coordinate dimension identifier for the closest pressure
+        level in model, equivalent in a sense to altitude
+        e.g. 'ilev', 'lev' for TIEGCM
+    scale : float
+        Scalar used to roughly translate a change in altitude with a
+        change in pressure level, the scale height.
+    obs_out : string
+        Label used to add interpolated altitudes into pbs Instrument
+        object.
+    tol : float
+        Allowed difference between observed and modeled altitudes.
+    autoscale : bool
+        if True, use 'units' in meta to determine relative scaling
+        between obs and mod
+
+    """
+
+    # Ensure the coordinate and data variable names are array-like
+    obs_coords = np.asarray(obs_coords)
+
+    # create initial fake regular grid index in inst
+    obs[mod_coord] = 0
+
+    # we need to create altitude index from model
+    # collect relevant inputs
+    # First, model locations for interpolation
+    # we use the dimensions associated with model altitude
+    # in the order provided
+    points = [mod.data.coords[dim].values if dim != 'time' else
+              mod.data.coords[dim].values.astype(int)
+              for dim in mod[mod_alt].dims]
+    # grab units from meta
+    mod_units = [mod.meta[dim]['units'] for dim in mod[mod_alt].dims]
+    # drop time units
+    mod_units = mod_units[1:]
+
+    # Determine the unit scaling between model and instrument data
+    inst_scale = np.ones(shape=len(obs_coords), dtype=float)
+    if autoscale:
+        for i, iname in enumerate(obs_coords):
+            if iname not in obs.variables:
+                raise ValueError(''.join(['Unknown instrument location index ',
+                                        '{:}'.format(iname)]))
+            if iname == obs_alt:
+                # pull out altitude units
+                inst_scale[i] = pyutils.scale_units(obs.meta[iname]['units'],
```

Comment:
Is this the preferred way to access something from metadata?  We should probably talk about this at the next pysat telecon.  Right now there are several ways to do this.

Reply:
There are several ways to get to the Metadata. One is built in (first), thee rest come along for the ride because of Pandas. Happy to talk about this on the next telecon. Note, I don't normally use the built-in method, so I'm definitely guilty of standards non-compliance. The first line below is the only one that works for multiple case standards.

ivm.meta['Sun_Status', 'desc']
Out[48]: u's/c Sun/Shadow Status Code'

ivm.meta['Sun_Status']['desc']
Out[49]: u's/c Sun/Shadow Status Code'

ivm.meta['Sun_Status'].desc
Out[50]: u's/c Sun/Shadow Status Code'

Mixed Case:

ivm.meta['Sun_Status', 'Desc']
Out[51]: u's/c Sun/Shadow Status Code'

ivm.meta['Sun_Status']['Desc']

KeyErrorTraceback (most recent call last)
<ipython-input-52-b3ecdf5308b3> in <module>()
----> 1 ivm.meta['Sun_Status']['Desc']

/Users/russellstoneback/Library/Enthought/Canopy/edm/envs/Python_2_again/lib/python2.7/site-packages/pandas/core/indexes/base.pyc in get_value(self, series, key)
   4386                     raise InvalidIndexError(key)
   4387                 else:
-> 4388                     raise e1
   4389             except Exception:  # pragma: no cover
   4390                 raise e1
KeyError: 'Desc' 

ivm.meta['Sun_Status'].Desc

AttributeErrorTraceback (most recent call last)
<ipython-input-53-9d747d71b59a> in <module>()
----> 1 ivm.meta['Sun_Status'].Desc

/Users/russellstoneback/Library/Enthought/Canopy/edm/envs/Python_2_again/lib/python2.7/site-packages/pandas/core/generic.pyc in __getattr__(self, name)
   5065             if self._info_axis._can_hold_identifiers_and_holds_name(name):
   5066                 return self[name]
-> 5067             return object.__getattribute__(self, name)
   5068 
   5069     def __setattr__(self, name, value):
AttributeError: 'Series' object has no attribute 'Desc'

Reply:
Most generally correct version:

ivm.meta['Sun_Status', ivm.meta.desc_label]

though, when using the labels, following should also work:
ivm.meta['Sun_Status'][ivm.meta.desc_label]

Reply:
We really need to pare this down so that there's just one way to do things :(

Reply:
I agree with the sentiment but I don't think we can prevent some of these options.

Meta itself provides the 'one' way to get all metadata about a variable,
ivm.meta[var_name]
which returns a collection. In this case, a Pandas collection, but the following is true of a dict as well. Getting a single metadata value out is
ivm.meta[var_name, ivm.*_label]
That is what Meta supports, but the user can do things we can't stop.
Such as,
a = ivm.meta[var_name]
b = a[sub_name]
since a is a collection. Also the same as 
b = ivm.meta[var_name][sub_name]
Because it is a Pandas collection, pandas itself also supports the dot notation as well,
b = ivm.meta[var_name].sub_name

Chained data access. What can you do....
We can work on documentation and our standards. That may be it though.